### PR TITLE
gdbus_codegen: Support --c-generate-autocleanup

### PR DESCRIPTION
**Pull request by robert-ancell**

### CODE REVIEW

**Comment by TingPing** [comment]
Maybe this should print a useful error on older versions of glib, perhaps even default to on with new glib?

**Comment by robert-ancell** [comment]
I think if it can be done safely this should be on by default.

**Comment by nirbheek** [comment]
What exactly does this flag generate? Does it emit code that uses the g_autoptr framework in glib?

**Comment by rilian-la-te** [comment]
Yes, it is)

**Comment by TingPing** [comment]
> This is not cross-platform compatible.

The macros are no-ops on unsupported compilers, it should be safe to always generate them here.

**Comment by nirbheek** [comment]
> The macros are no-ops on unsupported compilers, it should be safe to always generate them here.

Hum, you're right, it only uses `G_DEFINE_AUTOPTR_CLEANUP_FUNC` from what I can tell. The option was added because it broke some projects:

https://bugzilla.gnome.org/show_bug.cgi?id=763379#c14

The discussion on that bug is really long, and I reached the 40th comment before I gave up. But I think it's ok to enable it by default, because projects that break because of this can just disable it.

This option was added in 2.49.1, and the oldest version of glib that the meson module will work reliably with is 2.54, so it should be fine to add this as-is.

File: mesonbuild/modules/gnome.py [hunk 1]
```diff
@@ -890,6 +890,8 @@ def gdbus_codegen(self, state, args, kwargs):
             cmd += ['--c-namespace', kwargs.pop('namespace')]
         if kwargs.get('object_manager', False):
             cmd += ['--c-generate-object-manager']
+        if 'autocleanup' in kwargs:
+            cmd += ['--c-generate-autocleanup', kwargs.pop('autocleanup')]
```
Comment (nirbheek) [inline mesonbuild/modules/gnome.py]: Please add an `else` here which will add `['--c-generate-autocleanup', 'objects']` to `cmd`. We should turn it on by default. Also:

1. Please add a release note snippet, see `docs/markdown/snippets` for examples
2. Update the Gnome-module doc to say that the values can be `none`, `objects`, `all`, and that the default is `objects`
3. Validate the value passed via the kwarg and error out if it is invalid

**Comment by robert-ancell** [comment]
I've added a snippets file for both features, let me know if it's the wrong format.

**Pull request merged**